### ISSUE TITLE
Rules are automatically deleted

### ISSUE THREAD

**Issue by henshin** [FirefoxBar/HeaderEditor#91]
I'm not sure what's triggering this, but from time to time the rules are deleted and I have to import them again. 
A similar issue was reported previously but was regarding the groups being deleted. The groups are not deleted in my case, just the rules. However the reason might be related.
Not sure how can I provide more info, please let me know if you need anything else.

**Comment by sylingd** [comment]
Have you ever cleared cookies?

**Comment by henshin** [comment]
Maybe just for a specific website but never globally in Firefox. I'm not sure that this could be related to cookies.

**Comment by sylingd** [comment]
Rules are stored at IndexedDB. Behaviors known to cause the loss of IndexedDB data include:

* Clear cookies of this extension (In Firefox, IndexedDB is very close to cookies)

* Clear site storage of this extension

* An abnormal shutdown may theoretically cause database corruption, but no one has reported this problem still now

Previously, groups was stored at localStorage, now, groups is stored with storage API. The storage API is stored in different places than the IndexedDB.

**Comment by henshin** [comment]
I've just made a very simple test that reproduces the issue.
Imported my rules via file, saved settings, closed add-on settings, closed Firefox, started Firefox again and observe that the rules are missing.
I'm only using a single rule which uses a custom function, not sure if that matters.

**Comment by Thorin-Oakenpants** [comment]
- Do you have any other addons? i.e. can you reproduce in a vanilla profile?
- What is your FF version and release channel?
- Are you in Private Browsing mode? IDB is not available in PB mode
- What are your settings for "history": Options> Privacy & Security> History> Firefox will... (if custom what are the options under that)
- Do you clear anything on close?

**Comment by henshin** [comment]
Answering your questions:
- These are the other addons I have: Adblock, Disconnect, Proxy SwitchyOmega and TamperMonkey. However I tested disabling all the addons except HeadersEditor to test this and the behaviour is still the same.
- Using the latest available version of Firefox: 62.0.3 (64-bit)
- Not using private mode
- Firefox is set to "Remember history"
- Tracking protection is On
- Not clearing anything when closing the browser

I did noticed a strange behaviour though. Normally the addons have an token ID associated to them when we access the addon options, for example: moz-extension://af456517-1fe5-447c-aefb-b519c6207612/manage.html
For all the addons I have, these IDs are persistent across firefox sessions, except for the HeadersEditor. Everytime I close Firefox and re-open it, the ID for it changes. Kind of seems that the add on is reinstalled from scratch. Have you experienced this before?

**Comment by Thorin-Oakenpants** [comment]
Nothing from your setup looks out of place but I think you've narrowed the problem down with the UUID

> these IDs are persistent across firefox sessions, except for the HeadersEditor

The internal UUID's are random and assigned when installed - you can see them listed under about:debugging for quick reference. I've never seen that happen before (a new UUID each session).

Something to test
- New profile
- Install Header Editor & add a rule
- Look in your profile/storage/default folder, you should have a `moz-extension+++UUID` folder (since you only have one extension this should be the HE one - check it matches in about:debugging if you want)
- Close firefox
- Did the `moz-extension+++UUID` persist?

Something to try
- There is a pref in about:config you could *try* which will not fix the cause, but *may* stop the symptom (as your rules are kept in your profile folder under it's UUID)
- `extensions.webextensions.keepUuidOnUninstall` - default is false, change it to true
- do tests

However, we should really find the cause. In your real profile, go to about:debugging, open the HE manifest URL link, right click in an empty spot and select View Page Info. This will bring up a dialog, select the Permissions tab. I'm just guessing here, but there *may* be something that sticks out - but then nothing in here should make a difference AFAIK.

Also: dom.indexedDB.enabled should be default true (which I am sure it is, since you tested in a new profile, but just checkin' your real profile is the same)

**Comment by henshin** [comment]
So I've tried to create a new profile (while before I just tested using the guest one) and the issue was not present.
I figured it would be something to do with my profile folder so I tried to take a look at any possible issues.
I've found that inside my profile folder there were 9999 files named `prefs-xxxx.js` files all with 0 bytes. So I decided to close all my firefox tabs, closed firefox and deleted all of these files with 0 bytes. When I restarted Firefox I tried to install HE plugin again and the problem was gone. The UUID is now persistent. 
I'm not sure which of my actions solved the issue, but it is indeed solved :) Afterwards I also cleared any leftovers from previous HE `moz-extension+++UUID` "installations" from my storage\default

For the record, the Permissions tab had the default values and the dom.indexedDB.enabled was set to true as well.
I think that's it so thanks a lot for helping me debugging the issue @Thorin-Oakenpants!
